nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/Wait.java

```diff
@@ -0,0 +1,266 @@
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final ComponentLog logger = getLogger();
+
+        // cache key is computed from attribute 'RELEASE_SIGNAL_IDENTIFIER' with expression language support
+        final String cacheKey = context.getProperty(RELEASE_SIGNAL_IDENTIFIER).evaluateAttributeExpressions(flowFile).getValue();
+
+        // if the computed value is null, or empty, we transfer the flow file to failure relationship
+        if (StringUtils.isBlank(cacheKey)) {
+            logger.error("FlowFile {} has no attribute for given Release Signal Identifier", new Object[] {flowFile});
+            flowFile = session.penalize(flowFile);
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        // the cache client used to interact with the distributed cache
+        final DistributedMapCacheClient cache = context.getProperty(DISTRIBUTED_CACHE_SERVICE).asControllerService(DistributedMapCacheClient.class);
+
+        try {
+            // check for expiration
+            String waitStartTimestamp = flowFile.getAttribute(WAIT_START_TIMESTAMP);
+            if (waitStartTimestamp == null) {
+                waitStartTimestamp = String.valueOf(System.currentTimeMillis());
+                flowFile = session.putAttribute(flowFile, WAIT_START_TIMESTAMP, waitStartTimestamp);
+            }
+
+            long lWaitStartTimestamp = 0L;
+            try {
+                lWaitStartTimestamp = Long.parseLong(waitStartTimestamp);
+            } catch (NumberFormatException nfe) {
+                logger.error("{} has an invalid value '{}' on FlowFile {}", new Object[] {WAIT_START_TIMESTAMP, waitStartTimestamp, flowFile});
+                flowFile = session.penalize(flowFile);
+                session.transfer(flowFile, REL_FAILURE);
+                return;
+            }
+            long expirationDuration = context.getProperty(EXPIRATION_DURATION)
+                    .asTimePeriod(TimeUnit.MILLISECONDS);
+            long now = System.currentTimeMillis();
+            if (now > (lWaitStartTimestamp + expirationDuration)) {
+                logger.warn("FlowFile {} expired after {}ms", new Object[] {flowFile, (now - lWaitStartTimestamp)});
+                session.transfer(flowFile, REL_EXPIRED);
+                return;
+            }
+
+            // get notifying flow file attributes
+            Map<String, String> cachedAttributes = cache.get(cacheKey, keySerializer, valueDeserializer);
+
+            if (cachedAttributes == null) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("No release signal yet for {} on FlowFile {}", new Object[] {cacheKey, flowFile});
+                }
+                session.transfer(flowFile, REL_WAIT);
```

Comment:
My thought was that you'd configure the Schedule duration according to how fast you wanted it to be retried.  Do you think there should be an additional restriction?